src/rdsk.c

```diff
@@ -671,171 +1082,350 @@ int cache_device_unmap(struct RC_PROFILE *prof, unsigned char *string)
+		memset(status, 0x0, sizeof(status));
+		if (rd_prof->lock_status == TRUE) {
+			sprintf(status, "Locked");
+		} else if (rd_prof->lock_status == FALSE) {
+			sprintf(status, "Unlocked");
+		} else {
+			sprintf(status, "Unavailable");
 		}
+
+		printf(" RapidDisk Device %d: %s\tSize (KB): %llu\tUsage (KB): %llu\tStatus: %s\n", num,
+			   rd_prof->device, (rd_prof->size / 1024), (rd_prof->usage / 1024), status);
+		num++;
 		rd_prof = rd_prof->next;
 	}
+	printf("\nList of RapidDisk-Cache mapping(s):\n\n");
+	if (rc_prof == NULL) {
+		printf("  None\n");
+		goto list_out;
+	}
+	num = 1;
+	while (rc_prof != NULL) {
+		if (strstr(rc_prof->device, "rc-wb") != NULL) {
+			printf(" dm-writecache Target   %d: %s\tCache: %s  Target: %s (WRITEBACK)\n",
+				   num, rc_prof->device, rc_prof->cache, rc_prof->source);
+		} else {
+			printf(" RapidDisk-Cache Target %d: %s\tCache: %s  Target: %s (%s)\n",
+				   num, rc_prof->device, rc_prof->cache, rc_prof->source,
+				   (strncmp(rc_prof->device, "rc-wt_", 5) == 0) ? \
+				"WRITE THROUGH" : "WRITE AROUND");
+		}
+		num++;
+		rc_prof = rc_prof->next;
+	}
+	list_out:
+	printf("\n");
+	return SUCCESS;
+}
+
+/**
+ * It prints out the statistics of the wb cache device
+ *
+ * @param rc_prof the RC_PROFILE structure that contains the cache device name and the cache device size.
+ * @param cache the name of the cache device
+ *
+ * @return The function status result
+ */
+int cache_wb_device_stat(struct RC_PROFILE *rc_prof, char *cache)
+{
 
+	int rc = INVALID_VALUE;
+	char *msg;
+
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
+		print_error("%s", NULL, msg);
+//		printf("  No RapidDisk-Cache Mappings exist.\n\n");
+		return rc;
+	}
+	while (rc_prof != NULL) {
+		if (strcmp(cache, rc_prof->device) == SUCCESS)
+			rc = SUCCESS;
+		rc_prof = rc_prof->next;
+	}
 	if (rc != SUCCESS) {
-		printf("Error. Device %s does not exist.\n", string);
+		msg = "Error. Cache target %s does not exist";
+		print_error(msg, NULL, cache);
 		return -ENOENT;
 	}
+	struct WC_STATS *wc_stats = dm_get_status(cache, WRITEBACK);
+	int i = 2;
+	unsigned int *pvalue = NULL;
+	for (; i < (((sizeof(WC_STATS_OFFSETS) / sizeof(size_t)) - 1)); i++) {
+		pvalue = (unsigned int *) (((char *) wc_stats) + WC_STATS_OFFSETS[i]);
+		printf("%d ", *pvalue);
+	}
+	pvalue = (unsigned int *) (((char *) wc_stats) + WC_STATS_OFFSETS[i]);
+	printf("%d\n", *pvalue);
+	free(wc_stats);
+	return SUCCESS;
+}
 
-	sprintf(file, "/dev/%s", string);
+/**
+ * It prints the statistics of a cache device
+ *
+ * @param rc_prof This is the list of RapidDisk-Cache mappings.
+ * @param cache The cache device name.
+ *
+ * @return The function status result
+ */
+int cache_device_stat(struct RC_PROFILE *rc_prof, char *cache)
+{
+	int rc = INVALID_VALUE;
+	char *msg;
 
-	if ((fd = open(file, O_WRONLY)) < SUCCESS) {
-		printf("%s: open: %s\n", __func__, strerror(errno));
+	if (rc_prof == NULL) {
+		msg = "No RapidDisk-Cache Mappings exist.";
```

Comment:
I wonder if we should just predefine these strings in main.h instead of constantly setting msg. It would also reduce extra lines of code in this file.

Reply:
I did it in a new forthcoming commit, not pushed since not ready. I created `#define`s for many if not all of the error messages specified as static strings, but the work is not finished yet!